Document how to use demo project in `make run` (#722)
kedro-org/kedro-viz: CONTRIBUTING.md

### JavaScript development

> _*Note*: We use the [latest release of v14 Node.js](https://nodejs.org/download/release/latest-v14.x/) in our development environment._

Author: Maybe I don't know what I'm talking about here, but there seem to be many different version of 14.x.y and the exact one does seem to make a difference to the package-lock.json. So I thought best to fix on one particular version.

Reviewer: From what I know as long as it's v14 it would work for compiling our app, but I agree that specifying the latest version would help align the package-lock file. 

```suggestion
_*Note*: We suggest using the latest version of [Node.js v.14 (<=14.18.0)](https://nodejs.org/download/release/latest-v14.x/) in our development environment._
```

Author: Ah, that makes sense. Thanks for explaining @studioswong

Author: But why should we be suggesting <=14.18.0 rather than the latest 14.18.3 (which is what I've been using and seems to work ok)?

Reviewer: yea good point - yea let's go for <=4.18.3

```bash
DATA=random npm start
```

These are the supported dataset identifiers:

| Identifier | Data source |
|------------|-------------|
| `json` (default) | `/public/api/main` |
| `random` | Randomly-generated data |
| `demo` | `/src/utils/data/demo.mock.js` |
| `spaceflights` | `/src/utils/data/spaceflights.mock.json` |

By default in production, the app asynchronously loads JSON from the `/api/main` endpoint. You can replicate this in development by placing a JSON dataset in `/public/api/main`, using `main` as the name of the file, [without an extension](https://www.computerhope.com/issues/ch002089.htm). Note that operating systems often add hidden file extensions, so you might need to use a CLI to confirm the filename.

Alternatively, you can synchronously load one of the mock datasets in `/src/utils/data`. The 'spaceflights' dataset is mainly used as mock data for unit testing.

> Finally, you can use pseudo-random data, which is procedurally-generated on page load, and is often useful for local development. Random data can be seeded with a hash string, which will allow you to replicate a generated layout. You can supply a seed with a `seed` query string in the URL, e.g. `http://localhost:4141/?data=random&seed=oM4xauN4Whyse`. If you do not supply a seed, the app will generate a new pseudo-random one every time, and will output it to the browser console in case you wish to reuse it.

Reviewer: Nice one - thanks for adding this to bridge the gap! Great one. 

```suggestion
Finally, you can use pseudo-random data, which is procedurally-generated on page load, and is often useful for local development. Random data can be seeded with a hash string, which will allow you to replicate a generated layout. You can supply a seed with a `seed` query string in the URL, e.g. `http://localhost:4141/?data=random&seed=oM4xauN4Whyse`. If you do not supply a seed query to the URL, the app will generate a new pseudo-random seed on every browser refresh, and will output it to the browser console in case you wish to reuse it.
```

If you wish to point the server to a different Kedro project then you can do so by altering the `PROJECT_PATH` variable:

```bash
make run PROJECT_PATH=<path-to-your-test-project>/new-kedro-project
```

> **Note**: Once the development server is launched at port 4142, the local app will always pull data from that server. To prevent this, you can comment out the proxy setting in `package.json` and restart the dev server at port 4141.

#### Launch the development server with the `SQLiteSessionStore`

Kedro-Viz provides a `SQLiteSessionStore` that users can use in their project to enable experiment tracking functionality. If you want to use this session store with the development server, make sure you don't use a relative path when specifying the store's location in `settings.py`. For example, `demo-project` specifies the local `data` directory within a project as the session store's location as follows: 
```python
from kedro_viz.integrations.kedro.sqlite_store import SQLiteStore
SESSION_STORE_ARGS = {"path": str(Path(__file__).parents[2] / "data")}
```

> Owing to this coupling between the project settings and Kedro-Viz, if you wish to execute any Kedro commands on `demo-project` (including `kedro run`), you will need to install the Kedro-Viz Python package. To install your local development version of the package, run:

Author: You are right that there's two options - either install a released version from PyPI (`pip install kedro-viz`) or install your local development version (the command given). However, when you do `make run` it will always use the local development version (since that sets `PYTHONPATH=$(shell pwd)/package`). So for consistency I think it's best to only use the local development version for the session store as well - otherwise your project is using a different version of kedro-viz from the what the backend server uses. I will clarify this 🙂

> Owing to this coupling between the project settings and Kedro-Viz, if you wish to execute any Kedro commands on `demo-project` (including `kedro run`), you will need to install the Kedro-Viz Python package. To install your local development version of the package, run:

Reviewer: sounds great - yea let's just scope this to only mention the installation of the local development version ( which is what this section is mainly about anyways). 👍

Before launching a development server with a real Kedro project, you'd need to have [Python](https://www.python.org/)(>=3.7, <3.9) installed. We strongly recommend setting up [conda](https://docs.conda.io/en/latest/) to manage your Python versions and virtual environments. You can visit Kedro's [guide to installing conda](https://kedro.readthedocs.io/en/latest/02_get_started/01_prerequisites.html#conda) for more information.

The Kedro-Viz repository comes with an example project in the `demo-project` folder. This is used on the [public demo](https://demo.kedro.org/). To use it in your development environment, you need to install both the Kedro-Viz dependencies and a minimal set of dependencies  for the demo project:
```bash
pip3 install -r package/test_requirements.txt
pip3 install -r demo-project/src/docker_requirements.txt
```

Now build the application with:

```bash
make build
```

As far as the development server for the backend is concerned, you only need to run `make build` once when you first setup the project. You can then launch the server with:

```bash
make run
```

> This command will launch a Kedro-Viz server at [localhost:4142](http://localhost:4142) and serve data from `demo-project`. From then on, launching the app locally at [localhost:4141](http://localhost:4141) will pull data from the Kedro-Viz server that is running on port 4142.

Reviewer 2: I'm not sure how clear it is at this stage of the instructions that the backend and frontend are now running in tandem and that one is feeding the other.

Perhaps adding in another reference here to running the frontend with `npm start` will hammer home the point that now the backend and frontend are running together?

Do we need to be as explicit as saying something like, "You need to have two terminal windows open..." etc etc?

```suggestion
This command will launch a Kedro-Viz server at [localhost:4142](http://localhost:4142) and serve data from `demo-project`. From then on, launching the app locally with `npm start` at [localhost:4141](http://localhost:4141) will pull data from the Kedro-Viz server that is running on port 4142.
```

Author: This is a great point, thank you! I've made it clearer in 99c173a. Note that if you're just doing backend development, you can actually run the backend server at 4242 without the frontend one

Reviewer 2: Great! Thanks for adding.
